046-analyse-broker-metrics-before-rolling-kafka.md

```diff
@@ -0,0 +1,63 @@
+<!-- This template is provided as an example with sections you may wish to comment on with respect to your proposal. Add or remove sections as required to best articulate the proposal. -->
+
+# Analyse broker metrics before rolling kafka
+
+This proposal describes a solution for KafkaRoller's shortcoming with brokers in log recovery. 
+
+## Current situation
+
+The current logic in Kafka Roller:
+- Take a list of kafka pods and reorder them based on the readiness. 
+- Roll unready pods first without considering why it could be unready.
+- Before restarting a kafka pod, consider if it is a controller or it has an impact on the availability.
+- If a kafka pod does not become ready within the operational timeout (300000ms by default) after restarting, then force restart it without considering why it's taking a long time.
+
+A Kafka broker can take a long time to become ready while performing log recovery. In this case, KafkaRoller ends up force restarting the broker and this can continue indefinetly, because of the log recovery on the broker startup again. KafkaRoller currently does not have a way to know what state a broker is in before making a decision to force restart it.
+
+KafkaAgent currently collects `kafka.server:type=KafkaServer,name=BrokerState` metric and creates a `kafka-ready` file on disk if the broker state value is `3`(`RUNNING`). The Kafka readiness check passes if it finds the file on disk.
+
+## Motivation 
+
+To address the issue described above, the KafkaRoller needs to be aware of the current broker state, so that if a broker is performing log recovery, then it waits longer instead of force restarting it.
+
+## Proposal
+
+The proposal is a REST API endpoint on the KafkaAgent which exposes the BrokerState metric that it already uses. And modify KafkaRoller so that it queries the new endpoint to find out what state a broker is in. 
+
+1. Add a web server to KafkaAgent that exposes an endpoint which returns a value of metric
+    - Configure Jetty server to only accept SSL connections on a port 8443.
+    - Expose an endpoint `/broker-state`.
+    - Add a servlet to handle a `GET` request and return an integer value of `BrokerState` metric.
+    - It will reuse broker's own certificate, which will already be in the container.
+2. Update KafkaRoller to query the broker state from the KafkaAgent API
+    - Currently KafkaRoller obtains headless service for a given pod and uses it to boostrap Kafka AdminClient. It can use the same mechanism to get the address for the KafkaAgent endpoint.
+    - Instead of checking broker's readiness through `isReady()` function, utilise the new restful endpoint to determine if the broker is ready.
+    - After restarting a broker if it does not become ready within the operational timeout period, check whether the broker is performing log recovery by querying the new endpoint:
+        - If log recovery is in progress, do not restart the broker.
```

Comment:
If the operator detects that log recovery is in progress, it should probably issue some warning as well. Since this is a situation which should be ideally prevented.

Reply:
Since we now have log recovery metrics (KIP-831), it would be good to use them in a periodic info message and to check if any progress is being made.

Reply:
We could periodically check the recovery metrics using the existing pollingIntervalMs (1000 by default), however that might be too often unnecessarily. 

Or we could include the value of the recovery metrics in the warning message itself so on every retry (5 minutes or more), we will be able to see the progress. 

This also means we would have to expose log recovery metrics via KafkaAgent too?

Reply:
At least for my comment, this is more something I would expect the cluster operator to do when it sees the recovery happening in the KafkaRoller.

Reply:
I think we want a `WARN` log the first time we find we've exceeded the operation timeout, and a 2nd `WARN` message (including the total observed time from pod deletion/recreation to `RUNNING` state) once we observed the transition to `RUNNING`. This:

1. avoids spamming the logs
2. makes it easy to see that you are (or were) in this state
3. makes it easy to know how to what to reconfigure the timeout value to to avoid this in the future.

Reply:
I have addressed the suggestions in the latest commit. 

@tombentley do you think it's worth exposing the observed time as a metric from the operator? So that when we want to take it into consideration for changing the timeout, it could be more useful to see the observed value over time, rather than looking at one event.  But this could be for future improvements.

Reply:
Another idea is instead of logging a `WARN` and having to check if it's the first time exceeding the operation timeout, we could create a new error cause that says something like log recovery in progress. I noticed that reconciliation logging is quite noisy already and it does log error cause when the reconciliation fails.